Geometry/HcalTowerAlgo/src/HcalDDDGeometry.cc

```diff
@@ -200,70 +200,66 @@ HcalDDDGeometry::newCellImpl( const GlobalPoint& f1 ,
 
 void
 HcalDDDGeometry::newCell( const GlobalPoint& f1 ,
-              const GlobalPoint& f2 ,
-              const GlobalPoint& f3 ,
-              const CCGFloat*    parm ,
-              const DetId&       detId   )
+			  const GlobalPoint& f2 ,
+			  const GlobalPoint& f3 ,
+			  const CCGFloat*    parm ,
+			  const DetId&       detId   )
 {
   newCellImpl(f1,f2,f3,parm,detId);
   addValidID( detId );
 }
 
 void
 HcalDDDGeometry::newCellFast( const GlobalPoint& f1 ,
-              const GlobalPoint& f2 ,
-              const GlobalPoint& f3 ,
-              const CCGFloat*    parm ,
-              const DetId&       detId   )
+			      const GlobalPoint& f2 ,
+			      const GlobalPoint& f3 ,
+			      const CCGFloat*    parm ,
+			      const DetId&       detId   )
 {
   newCellImpl(f1,f2,f3,parm,detId);
   m_validIds.emplace_back(detId);
 }
 
-const CaloCellGeometry* 
-HcalDDDGeometry::cellGeomPtr( uint32_t din ) const
-{
-  const CaloCellGeometry* cell ( nullptr ) ;
-  if( m_hbCellVec.size() > din )
-  {
-    cell = &m_hbCellVec[ din ] ;
+const CaloCellGeometry* HcalDDDGeometry::getGeometryRawPtr (const DetId& id) const {
+  const unsigned int din = CaloGenericDetId(id).denseIndex();
+  const CaloCellGeometry* cell(nullptr);
+  if (m_hbCellVec.size() > din) {
+    cell = (&m_hbCellVec[din]);
+  } else if (m_hbCellVec.size()+m_heCellVec.size() > din) {
+    const unsigned int ind (din - m_hbCellVec.size() ) ;
+    cell = (&m_heCellVec[ind]);
+  } else if (m_hbCellVec.size()+m_heCellVec.size()+m_hoCellVec.size() > din) {
+    const unsigned int ind (din - m_hbCellVec.size() - m_heCellVec.size());
+    cell = (&m_hoCellVec[ind]);
+  } else if (m_hbCellVec.size()+m_heCellVec.size()+m_hoCellVec.size()+
+	     m_hfCellVec.size() > din) {
+    const unsigned int ind (din - m_hbCellVec.size() - m_heCellVec.size() -
+			    m_hoCellVec.size() ) ;
+    cell = (&m_hfCellVec[ind]);
   }
-  else
-  {
-    if( m_hbCellVec.size() +
-	m_heCellVec.size() > din )
-    {
-      const unsigned int index ( din - m_hbCellVec.size() ) ;
-      cell = &m_heCellVec[ index ] ;
-    }
-    else
-    {
-      if( m_hbCellVec.size() +
-	  m_heCellVec.size() +
-	  m_hoCellVec.size() > din )
-      {
-	const unsigned int index ( din 
-				   - m_hbCellVec.size() 
-				   - m_heCellVec.size() ) ;
-	cell = &m_hoCellVec[ index ] ;
-      }
-      else
-      {
-	if( m_hbCellVec.size() +
-	    m_heCellVec.size() +
-	    m_hoCellVec.size() +
-	    m_hfCellVec.size() > din )
-	{
-	  const unsigned int index ( din 
-				     - m_hbCellVec.size() 
-				     - m_heCellVec.size() 
-				     - m_hoCellVec.size() ) ;
-	  cell = &m_hfCellVec[ index ] ;
-	}
-      }
-    }
+  
+  return (( nullptr == cell || nullptr == cell->param()) ? nullptr : cell ) ;
+}
+
+std::shared_ptr<const CaloCellGeometry> HcalDDDGeometry::cellGeomPtr( unsigned int din ) const {
+  static const auto do_not_delete = [](const void*){};
```

Comment:
when I proposed getGeometryRawPtr, it seemed to also be easier to have this method implemented as
```C++
auto ptr = getGeometryRawPtr(din);
static const auto do_not_delete = [](const void*){};
return ptr == nullptr ? nullptr : std::shared_ptr<const CaloCellGeometry>(ptr, do_not_delete);
```